object-store/src/main/java/org/hypertrace/config/objectstore/ConfigObjectScopeResolver.java

```diff
@@ -0,0 +1,131 @@
+package org.hypertrace.config.objectstore;
+
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Message;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Resolver class to resolve configs based on scope
+ *
+ * @param <T> Raw config type
+ * @param <U> Resolved config type
+ * @param <S> Scope
+ */
+public abstract class ConfigObjectScopeResolver<T extends Message, U extends Message, S> {
+
+  protected final Optional<U> defaultConfig;
+
+  public ConfigObjectScopeResolver(U defaultConfig) {
+    this.defaultConfig = Optional.of(defaultConfig);
+  }
+
+  public ConfigObjectScopeResolver() {
+    this.defaultConfig = Optional.empty();
+  }
+
+  public List<U> getAllResolvedConfigData(List<T> configDataList) {
+    Map<S, T> configDataMap = getConfigDataMap(configDataList);
+    return configDataList.stream()
+        .flatMap(configData -> resolveConfig(extractScope(configData), configDataMap).stream())
+        .collect(Collectors.toUnmodifiableList());
+  }
+
+  public Optional<U> getResolvedData(List<T> configDataList, S scope) {
+    Map<S, T> configDataMap = getConfigDataMap(configDataList);
+    return resolveConfig(scope, configDataMap);
+  }
+
+  /**
+   * Method to extract scope object from given config object
+   *
+   * @param configData
+   * @return Scope object
+   */
+  protected abstract S extractScope(T configData);
+
+  /**
+   * Method to return scopes with which the given scope needs to be resolved in increasing order of
+   * priority For example, API scope would return a list {customer, env, service, api}
+   *
+   * @param scopeObject
+   * @return list of associated scopes in increasing order of priority
+   */
+  protected abstract List<S> getResolutionScopesWithIncreasingPriority(S scopeObject);
+
+  /**
+   * Method to return object of resolved config type from object store config data
+   *
+   * @param configStoreData
+   * @return
+   */
+  protected abstract U convertConfig(T configStoreData);
+
+  /**
+   * Method to resolve config store object with object of resolved config type
+   *
+   * @param fallbackConfig
+   * @param preferredConfig
+   * @return
+   */
+  protected abstract U resolveConfig(U fallbackConfig, T preferredConfig);
+
+  private Map<S, T> getConfigDataMap(List<T> configDataList) {
+    return configDataList.stream()
+        .collect(Collectors.toMap(this::extractScope, Function.identity()));
+  }
+
+  private Optional<U> resolveConfig(S scope, Map<S, T> configDataMap) {
+    Optional<T> mergedConfig =
+        getResolutionScopesWithIncreasingPriority(scope).stream()
+            .flatMap(scopeObject -> Optional.ofNullable(configDataMap.get(scope)).stream())
+            .reduce(
+                (fallbackConfig, preferredConfig) ->
+                    (T) mergeConfigs(fallbackConfig, preferredConfig));
+    return mergedConfig
```

Comment:
It feels a bit strange to merge the stored configs via algo, then ask the implementer to implement how to merge the merged + default configs. What do you think of removing `protected abstract U resolveConfig(U fallbackConfig, T preferredConfig);` and converting to U before merging so we can do a single reduce via generic merge? something like
```
Stream.concat(
  defaultConfig.stream(), 
  getResolutionScopesWithIncreasingPriority(scope).stream().map(this::convertConfig)
).reduce(...)`

Reply:
no wouldn't work in many cases because in many cases, we store configs with optional values in config store while the resolved value returned would need all values present.. for example,

```
msg T {
 optional bool disabled = 1;
 optional bool blocking = 2;
 optional string category = 3;
}
msg U {
 bool disabled = 1;
 bool blocking = 2;
 string category = 3;
}
```

Now let's say we have configs stored as `T-api1={disabled=true}, T-service1={blocking=true}` and default config as `U-default={disabled=false, blocking=false, category=latent}`.. now we want resolved config as `U-api1-resolved={disabled=true, blocking=true, category=latent}`.. but if we convert them individually to U, we would get `U-api1={disabled=true, blocking=false, category=latent}, T-service1={disabled=false, blocking=true, category=latent}` and merging them would result in `U-api1-resolved={disabled=true, blocking=false, category=latent}` because now all values are set for api1, not taking the fallback values from service1..

1 idea I was thinking about was not having `protected abstract U resolveConfig(U fallbackConfig, T preferredConfig);` - only keeping the `convertConfig()` method and caller can choose to use/not use default config to convert.. but that might be misleading, in case they want to convert any `T` config to `U` config independently too (sometimes very useful).. so kept these 2 different..

Reply:
I think I'm following - definitely on the first part at least. The part I'm struggling with is - if I have to implement the accumulator(U + T -> U) function, what's the point of the out of the box merge? The accumulator could just as easily do the work. 

So if we want to use generic merging logic: 

We need to get each value in the merge chain to be the same type. Your explanation on why that shouldn't be the returned type U makes sense, so I would guess this would look like changing the default to be of type T, merging generically then using convert (T -> U) with the result. Throw if no values in merge chain (which means no default given).

If we don't and want to use an accumulator instead:

This is similar to what's in current PR. We need a starting point, so we require exactly one of [default (U) OR converter (T->U)] AND accumulator (U + T -> U). This case only throws if no default or values (no T to convert).  Basically, we don't need both ways of seeding the reduce, nor do we need the generic merge logic if we go this route, at least that's my understanding.

Reply:
This is complicating matters and might not solve the reqs.. changed the default config type to `T` - that might make more sense and simplifies the logic too..